### PR TITLE
Improved profiling support

### PR DESCRIPTION
**Pull request by hiker**

### CODE REVIEW

**Comment by hiker** [comment]
First try for #296 (Support for Dr Hook). It also adds a complete makefile set to compile all wrapper libraries.

One point I am not 100% sure about: where do you want to have the documentation about how to compile the various libraries (and which options might need to be defined). At this stage I have added details to the READMEs under lib/profiling, and only a short, generic statement to the docs.

I would prefer not to duplicate the docs - i.e. have to have two files to keep in synch with the same info (README and profiling.rst). 

Also note that I can not install Dr Hook - the install target seems severely broken (it tries to install some .h files from an invalid path, and does not install .mod files at all). So my default relies on using the build directory.

File: lib/profiling/README [hunk 1]
```diff
@@ -31,9 +51,19 @@ dl_timer
 This is a wrapper library that maps the PSyclone profiling API
 to the dl_timer API. A copy of dl_timer can be downloaded from
 https://bitbucket.org/apeg/dl_timer
-It uses the ProfileData type and dl_timer's timer_register function
-to store the module/region name and the index used by dl_timer.
-This library is thread-safe.
+
+To compile the PSyclone wrapper library for dl_timer, one of the following
+two Makefile variables must be set:
+- DL_TIMER_ROOT This is the path to the apeg-dl_timer directory in which
```
Comment (arporter) [inline lib/profiling/README]: I've tested this and it works :-)

File: lib/profiling/README [hunk 2]
```diff
@@ -56,3 +86,53 @@ psy_time_step_mod:swlat_update_c    11  0.19000E+01   0.17272E+00  0.24E-02
 -----------------------------------------------------------------------------
 * corrected for systematic error
 =============================================================================
+
+
+Dr Hook
+-------
+This wrapper library interfaces with ECMWF's Dr Hook library. This library
+appears not to be available as open source on a public server. It provides more
+functionality than just profiling, see the Dr Hook documentations for details.
```
Comment (arporter) [inline lib/profiling/README]: s/documentations/documentation/ and is this documentation available publicly anywhere?

File: lib/profiling/README [hunk 3]
```diff
@@ -56,3 +86,53 @@ psy_time_step_mod:swlat_update_c    11  0.19000E+01   0.17272E+00  0.24E-02
...
+In the version tested (1.0.0) it appears that the installation target in the
+Dr Hook distribution is broken, so the following instructions and the default
+values assume that Dr Hook is comiled in a drhook subdirectory called 'build' and not
```
Comment (arporter) [inline lib/profiling/README]: s/comiled/compiled/

File: lib/profiling/README [hunk 4]
```diff
@@ -56,3 +86,53 @@ psy_time_step_mod:swlat_update_c    11  0.19000E+01   0.17272E+00  0.24E-02
...
+
+To compile the PSyclone wrapper library for Dr Hook, chose one of the following
+two options to specify the path to the Dr Hook installation:
+- DRHOOK_ROOT This is the path to the Dr Hook root directory in which
+  Dr Hook was compiled. It defaults to ../../../../drhook, which
+  means if Dr Hook is compiled next to the PSyclone root dir, it will be
+  found by default. This will set DRHOOK_INCLUDE to .../drhook/include
+  and DRHOOK_MODULES to .../drhook/build/module, so that the .mod files
+  created in the build processes are found.
+- DRHOOK_INCLUDE and DRHOOK_MODULES Setting those environment variables
```
Comment (arporter) [inline lib/profiling/README]: s/those/these/ and s/explicitely/explicitly/ below.

File: lib/profiling/README [hunk 5]
```diff
@@ -56,3 +86,53 @@ psy_time_step_mod:swlat_update_c    11  0.19000E+01   0.17272E+00  0.24E-02
...
+  explicitely will allow you to flexibly point to an existing Dr Hook installation.
+  It is not known if a proper Dr Hook installation will install include
```
Comment (arporter) [inline lib/profiling/README]: s/if/whether/

File: lib/profiling/dl_timer/Makefile [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Assume that apeg-dl_timer is 'next' to psyclone
```
Comment (arporter) [inline lib/profiling/dl_timer/Makefile]: Please add the license notice to this file.

File: lib/profiling/drhook/Makefile [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# Assume that apeg-dl_timer is 'next' to psyclone
```
Comment (arporter) [inline lib/profiling/drhook/Makefile]: License info please.

File: lib/profiling/drhook/drhook.f90 [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors J. Henrichs, Bureau of Meteorology
+
+
+!> An implemention of profile_mod which wraps the use of the Dr Hook.
```
Comment (arporter) [inline lib/profiling/drhook/drhook.f90]: s/of the Dr Hook/of Dr Hook/

File: lib/profiling/drhook/drhook.f90 [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  ! Parameters: 
+  ! module_name:  Name of the module in which the region is
+  ! region_name:  Name of the region (could be name of an invoke, or
+  !               subroutine name).
+  ! profile_data: Persistent data used by the profiling library.
+  subroutine ProfileStart(module_name, region_name, profile_data)
+    use yomhook, only : lhook, dr_hook
+    implicit none
+
+    character*(*) :: module_name, region_name
```
Comment (arporter) [inline lib/profiling/drhook/drhook.f90]: Please give explicit `intent` attributes.

File: lib/profiling/drhook/drhook.f90 [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+! Authors J. Henrichs, Bureau of Meteorology
+
+
+!> An implemention of profile_mod which wraps the use of the Dr Hook.
+
+module profile_mod
+
+  use parkind1, only : jprb
+  type :: ProfileData
+     real(kind=jprb) :: zhook_handle
```
Comment (arporter) [inline lib/profiling/drhook/drhook.f90]: Please briefly document each member of this type.

File: lib/profiling/drhook/drhook.f90 [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  ! ---------------------------------------------------------------------------
+  ! Ends a profiling area. It takes a ProfileData type that corresponds to
+  ! to the ProfileStart call.
+  ! profile_data: Persistent data used by the profiling library.
+  ! 
+  subroutine ProfileEnd(profile_data)
+    use yomhook, only : lhook, dr_hook
+    implicit none
+
+    type(ProfileData) :: profile_data
```
Comment (arporter) [inline lib/profiling/drhook/drhook.f90]: `intent(inout)`?

File: lib/profiling/simple_timing/Makefile [hunk 12]
```diff
@@ -0,0 +1,17 @@
+F90?=gfortran
```
Comment (arporter) [inline lib/profiling/simple_timing/Makefile]: License please.

File: lib/profiling/template/Makefile [hunk 13]
```diff
@@ -0,0 +1,12 @@
+F90?=gfortran
```
Comment (arporter) [inline lib/profiling/template/Makefile]: License please.

File: doc/profiling.rst [hunk 14]
```diff
@@ -220,9 +220,16 @@ As an example::
 Interface to Third Party Profiling Tools
```
Comment (arporter) [inline doc/profiling.rst]: Please also update the line in this file that says that "PSyclone currently supports dl_timer"

File: doc/profiling.rst [hunk 15]
```diff
@@ -220,9 +220,16 @@ As an example::
...
 ----------------------------------------
 PSyclone comes with wrapper libraries to support usage of
-dl_timer and a simple non-thread-safe timing
-library. Support for DrHook and TAU will be added in the
-future.
+Dr Hook, dl_timer and a simple non-thread-safe timing
+library. Support for further profiling libraries will be
+added in the future. To compile the wrapper libraries,
+change into the directory ``lib/profiling`` of PSyclone
+and type ``make`` to compile all wrapper. If only some
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/profiling.rst]: I can give you a copy of DrHook (which I received from UKMO) if you want to.

File: lib/profiling/README [hunk 16]
```diff
@@ -1,6 +1,26 @@
 This directory contains wrapper libraries that can be used with
 PSyclone's profiling interface.
 
+Compilation
+-----------
+The top level makefile can be used to compile some or all profiling libraries:
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/profiling/README]: I hope I got all instances of this fixed, well spotted.

File: lib/profiling/README (hunk 1 above)
Comment (hiker) [reply lib/profiling/README]: Hooray :)

File: lib/profiling/README (hunk 2 above)
Comment (hiker) [reply lib/profiling/README]: Nope, you need to know the secret handshake.

Seriously: I have no idea how to get DrHook. I got a copy from one of the UKMO people when we discussed that we should support DrHook, and the comment was (from memory) that DrHook is not available on any public repo. Googling found:
https://software.ecmwf.int/wiki/download/attachments/19661682/drhook.pdf
But I don't know if this is a link we should be using, trying to play around with the URL resulted only in 'page not found' (e.g. use just up to 'download' or so), and searching on the wiki for hook only resulted in various pdf files.

So .... no idea :(

**Comment by hiker** [comment]
OK, I think I've fixed everything. If you want to add the link to the Dr Hook pdf, let me know. Latest master merged in, ready for review again.

File: doc/profiling.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/profiling.rst]: I was tempted to duck this responsibility but I think that as reviewer I should probably take it on the chin :-) Please send me a tarball and I'll have a go. (Does it come with an example of how to use it?)

File: lib/profiling/README [hunk 17]
```diff
@@ -18,8 +19,8 @@ FFLAGS        -g                        Flags to use when compiling.
 
 Using
 make clean
-will clean all compiled libraries. You can also compile each library by
-changing into the corresponding directory and invoking 'make'.
+will clean all compiled library interfaces. You can also compile each library
+by interface changing into the corresponding directory and invoking 'make'.
```
Comment (arporter) [inline lib/profiling/README]: s/by interface/interface by/

**Comment by arporter** [comment]
There's just one minor typo left. I'm going to follow-up with ECMWF to see about whether there is any public documentation. I shall also test the Dr Hook profiling interface once @hiker has given me the tarball...

**Comment by TeranIvy** [comment]
Hi both! I am currently trying to enable PSyclone interface to DrHook library and it is not quite clear to me how this should work.  None of the files in `PSyclone/lib` repository directory is copied into PSyclone installation (as a Python package). Does it mean that I need to compile the wrapper to DrHook from a working copy of PSyclone repository and provide the path to our build system?
If so, this is a bit inconvenient. Is there any way that the compiled wrappers could be provided as part of the installation?

**Comment by arporter** [comment]
The profiling interface is not Python and therefore distutils doesn't know anything about it. I've done a quick google and nothing leaps out for using distutils to compile arbitrary code. Joerg has provided Makefiles so you would need to `git clone` and then build the relevant interface. We have debated breaking the Profiling code out into a separate repository but I don't think that would make your life any easier?

**Comment by TeranIvy** [comment]
I built the interface to DrHook by pointing `DRHOOK_ROOT`, `DRHOOK_INCLUDE` and `DRHOOK_MODULES` to the location of a Dr Hook library used by the UM people. I also added location of the compiled wrapper interface (`profile_mod`) to `LDFLAGS`. Is there anything else I need to do?

**Comment by hiker** [comment]
> I built the interface to DrHook by pointing `DRHOOK_ROOT`, `DRHOOK_INCLUDE` and `DRHOOK_MODULES` to the location of a Dr Hook library used by the UM people. I also added location of the compiled wrapper interface (`profile_mod`) to `LDFLAGS`. Is there anything else I need to do?
Yes, usage in reality might be a bit more awkward. Once you have the interface built, you should be ready to do. Use psyclone with:
```
--force-profile invokes --force-profile kernels 
```
(whatever you want/need - the 'force-' part is only required if you have other transformations happening, that could interfere with the transforms). You can check that this inserts the calls to the profiling library as expected. Then at link time make sure to link in the interface library for drhook, and drhook itself.

**Comment by hiker** [comment]
@TeranIvy and I fixed the issues with using Dr Hook with her code (the build environment put variables in an unexpected order), so this is ready now.

File: doc/profiling.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/profiling.rst]: @hiker and @teranivy have now done this.

File: lib/profiling/README (hunk 2 above)
Comment (arporter) [reply lib/profiling/README]: The documentation (and the package itself) are not public so I'm happy to leave this.

**Pull request merged**